moneybox/api/authentication.py

```diff
@@ -0,0 +1,18 @@
+from rest_framework import exceptions
+from rest_framework.authentication import BaseAuthentication
+
+from users.models import APIUser
+
+
+class CustomAuthentication(BaseAuthentication):
+    def authenticate(self, request):
+        if request.headers.get("Authorization") is not None:
+            auth_header = request.headers.get("Authorization")
+            # Auth header should look like "Token <uuid_string>"
+            token_from_header = auth_header.split(" ")[1]
+            if APIUser.objects.filter(token=token_from_header).exists():
+                authenticated_user = APIUser.objects.filter(token=token_from_header).first()
+                token = authenticated_user.token
+                return (authenticated_user, token)
```

Comment:
А нам точно токен нужно возвращать, если возвращаем самого юзера?

Reply:
Тут я выяснила, что если не происходит аутентификация, то в **request.user** потом всё равно можно обнаружить такой объект, как AnonymousUser. То есть это свойство всё равно будет присутствовать в реквесте, даже если это не будет наш юзер. Возвращая ещё и токен, я как бы гарантирую, что тогда уж точно это наш аутентифицировавшийся юзер.
В доке DRF я прочитала, что возвращаемый **request.auth** (это как раз второй элемент в возвращаемом кортеже) можно использовать потом в пермишене:
_The **request.auth** property is used for any additional authentication information, for example, it may be used to represent an authentication token that the request was signed with._
_Permission checks will typically use the authentication information in the request.user and **request.auth** properties to determine if the incoming request should be permitted._

Reply:
Согласен :)